### PR TITLE
Fixing errors with localStorage for Safari and Chrome 

### PR DESCRIPTION
**Pull request by nicar**
in return to: #545

### Description
Problems (see Screenshots below) with localStorage on Chrome and Safari (and Edge?) if MEW is opened offline
for Chrome: if third party cookies are disabled
for Safari: fuckshitstack in general

The fix resolves most of the problems and renders the site at least usable by checking for localStorage and otherwise falling back to a "polyfill" via cookies. 

### Note
Code works, but probably needs refactoring/relocating – need advise on that one though.

### Tested with

release `v3.9.8.5` and `v.3.9.9.1`

on OS X 10.10.5 (Yosemite) and MacOS 10.12 (Sierra)
Chrome: Version 59.0.3071.115 (Official Build) (64-bit)
Safari: Version 10.1.1 (10603.2.5)
Firefox: 54.0.1 (64-bit)

---

## Earlier problems

Release worked for

| Browser | file:// | localhost |
|-------|:-------:|:-----------:
|Firefox   |    yes       |      yes|
FF priv.mode  |   yes       |      yes
Chrome   |     x          |     yes
Chrome priv.mode   |  x        |       yes
Safari     |   x    |           yes
Safari priv.mode   |  x       |         x
Edge    |      ? | ?

__Chrome, from file://__
mew_chrome_file

---

**Safari, from file://**
mew_safari_file

---

**Safari, from localhost in private mode**
mew_safari_private_localhost

---

## Fix
The fix resolves most of the problems and renders the site at least usable by checking for localStorage and otherwise falling back to a "polyfill" via cookies. 

As far as I have tested

 Browser   |         file://  |   localhost
|-------|:-------:|:-----------:|
|Firefox   |    yes       |      yes|
FF priv.mode  |   yes       |      yes
Chrome   |     _limited_          |     yes
Chrome priv.mode   |  _limited_        |       yes
Safari     |   yes    |           yes
Safari priv.mode   |  _limited_       |         _limited_

_limited_ = No error, but settings may not be stored and lost on reload (Chrome) respective on close and reopen (Safari).

Chrome does NOT allow cookies on file:// if third party cookies are blocked.

User enabled solutions for Chrome: 
       allow third party cookies
    OR set exception for "file://"
    OR startup chrome with flag "--enable-file-cookies" (see https://bugs.chromium.org/p/chromium/issues/detail?id=535)

Code Sources:
https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API/Using_the_Web_Storage_API#Testing_for_availability
https://developer.mozilla.org/en-US/docs/Web/API/Storage/LocalStorage

### CODE REVIEW

File: app/scripts/globalFuncs.js [hunk 1]
```diff
@@ -227,14 +227,24 @@ globalFuncs.removeTokenFromLocal = function(symbol, tokenObj) {
...
+        cookieStorage: false,
+
         isAvailable: function() {
-            return typeof localStorage != "undefined";
+            // return typeof localStorage != "undefined";
+            
+            // console.log("isAvailable() globalFuncs.storageAvailable: " + globalFuncs.storageAvailable('localStorage'));
+            if(!globalFuncs.storageAvailable('localStorage') && this.cookieStorage == false) {
+                this.cookieStorage = true;
+                globalFuncs.createCookieLocalStorage();
```
Comment (crptm) [inline app/scripts/globalFuncs.js]: we should move this side-effect elsewhere, since setItem calls this

File: app/scripts/globalFuncs.js [hunk 2]
```diff
@@ -245,6 +255,99 @@ globalFuncs.localStorage = {
...
+            // acknowledge QuotaExceededError only if there's something already stored
+            storage.length !== 0;
+    }
+}
+
+/* Create fallback for localStorage based on cookies */
+globalFuncs.createCookieLocalStorage = function() {
+    // console.info("creating localStorage fallback");
+    if (!globalFuncs.localStorage.isAvailable()) {
+      Object.defineProperty(window, "localStorage", new (function () {
```
Comment (crptm) [inline app/scripts/globalFuncs.js]: Would this work in all browsers if localStorage is present but isnt working for some reason?

Can we move this (and available check) to separate file and polyfill once on load is needed?

File: app/scripts/globalFuncs.js [hunk 3]
```diff
@@ -245,6 +255,99 @@ globalFuncs.localStorage = {
...
+        Object.defineProperty(oStorage, "key", {
+          value: function (nKeyId) { return aKeys[nKeyId]; },
+          writable: false,
+          configurable: false,
+          enumerable: false
+        });
+        Object.defineProperty(oStorage, "setItem", {
+          value: function (sKey, sValue) {
+            if(!sKey) { return; }
+            document.cookie = escape(sKey) + "=" + escape(sValue) + "; expires=Tue, 19 Jan 2038 03:14:07 GMT; path=/";
```
Comment (crptm) [inline app/scripts/globalFuncs.js]: If cookie size allows i would prefer a single cookie with all data.

Comment (crptm) [reply app/scripts/globalFuncs.js]: Would `getItem` right after `setItem` work?

**Comment by tayvano** [comment]
@crptm Is this mergeable or needs adjustments?

File: app/scripts/globalFuncs.js (hunk 1 above)
Comment (nicar) [reply app/scripts/globalFuncs.js]: Sure, like commented – not very elegant yet. Any proposal? Needs to run at least once to build the cookie storage if localStorage is not available, probably after initialization/on load?

File: app/scripts/globalFuncs.js (hunk 2 above)
Comment (nicar) [reply app/scripts/globalFuncs.js]: > Would this work in all browsers if localStorage is present but isnt working for some reason?

Not sure if all browsers but it seems like all major ones. E.g. private browsing in Safari supports localStorage but doesn't make it available (0 quota). 

> Can we move this (and available check) to separate file and polyfill once on load is needed?

I can try ;) If you have any hints/tips I'll gladly take them in.

File: app/scripts/globalFuncs.js (hunk 3 above)
Comment (nicar) [reply app/scripts/globalFuncs.js]: > If cookie size allows i would prefer a single cookie with all data.

May I ask why? Serializing everything into one cookie doesn't seem like an advantage to me over "individual" key value cookies while these mimic the localStorage quite fine. When document.cookie outputs all of them as one string anyway. 

> Would getItem right after setItem work?

Can't see why it shouldn't.

File: app/scripts/globalFuncs.js (hunk 2 above)
Comment (crptm) [reply app/scripts/globalFuncs.js]: I mean specifically defining property on window over existing native prop. No reason it shouldnt work just curious if we can confirm that define property works for everything on window

File: app/scripts/globalFuncs.js (hunk 3 above)
Comment (crptm) [reply app/scripts/globalFuncs.js]: > Can't see why it shouldn't.

Getitems doesnt read from cookie, setitem doesnt write to this

Comment (crptm) [reply app/scripts/globalFuncs.js]: re:cookie it is not a big deal

File: app/scripts/globalFuncs.js (hunk 1 above)
Comment (crptm) [reply app/scripts/globalFuncs.js]: See below, we can make it a module that runs onload

**Comment by crptm** [comment]
We can merge if @nicar is busy. One thing i want to cleared up/fixed is getItem after setItem

File: app/scripts/globalFuncs.js (hunk 3 above)
Comment (nicar) [reply app/scripts/globalFuncs.js]: Examples (if I got you wrong let me know):

mew_cookies

mew_cookies_example

mew_cookies_examples

Comment (crptm) [reply app/scripts/globalFuncs.js]: @nicar is this your wrapper on screens?

File: app/scripts/main.js [hunk 4]
```diff
@@ -1,4 +1,5 @@
 'use strict';
+require('./localStoragePolyfill');
```
Comment (crptm) [inline app/scripts/main.js:2]: 👍

File: app/scripts/globalFuncs.js (hunk 3 above)
Comment (nicar) [reply app/scripts/globalFuncs.js]: It's the recent commit, console in Safari on localhost, private browsing …  just to illustrate the different situations and resulting cookies (set, get, remove on the different references to localStorage).

**Review by crptm — APPROVED** [review]
LGTM

**Pull request merged**